File: Algorithms/Dynamic Programming/python/house_robber.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        """Dynamic Programming solution for House Robber problem.

        Args:
            nums: Given a list of non-negative integers representing the amount of money of each house.

        Returns:
            Return the maximum amount of money without robbing two adjacent houses.

        Time Complexity: O(n)
        Space Complexity: O(1) 

        """
        if not nums:
            return 0
        pre_max = cur_max = 0
        for n in nums:
            pre_max, cur_max = cur_max, max(pre_max + n, cur_max)
        return cur_max

    def rob_circle(self, nums: List[int]) -> int:
        """Dynamic Programming solution for House Robber problem if the houses arranged in a circle.

        Args:
            nums: Given a list of non-negative integers representing the amount of money of each house.

        Returns:
            Return the maximum amount of money without robbing two adjacent houses.

        Time Complexity: O(n)
        Space Complexity: O(1) 

        """
        return nums[0] if len(nums) == 1 else max(self.rob(nums[1:]), self.rob(nums[:-1]))

File: Algorithms/Dynamic Programming/java/HouseRobber.java
/**
 * You are a professional robber planning to rob houses along a street. Each
 * house has a certain amount of money stashed, the only constraint stopping you
 * from robbing each of them is that adjacent houses have security system
 * connected and it will automatically contact the police if two adjacent houses
 * were broken into on the same night.
 * 
 * Given a list of non-negative integers representing the amount of money of
 * each house, determine the maximum amount of money you can rob tonight without
 * alerting the police.
 * 
 */
public class HouseRobber {
    /**
     * @param nums: an integer array without negative numbers.
     * @return the maximum amount of money follows the robbing rule.
     * 
     *         Time Complexity: O(n), Space Complexity: O(1)
     */
    public int rob(int[] nums) {
        if (nums.length == 0)
            return 0;
        int preMax = 0, curMax = 0;
        for (int n : nums) {
            int tmp = preMax;
            preMax = curMax;
            curMax = Math.max(tmp + n, curMax);
        }
        return curMax;
    }
}